### ISSUE TITLE
Composibiity, imports, reuse

### ISSUE THREAD

**Issue by zevv** [zevv/npeg#9]
Npeg would greatly benefit from a mechanism providing reuse of rules using some kind of import mechanism to pull in pre-defined patterns from libraries. 

With this mechanism Npeg could provide a library of predefined common patterns to be shipped with Npeg to simplify common operations like parsing of IPv4/IPv6 addresses, URIs, e-mail addresses.

I'm looking at an implementation but I seem to run into the limits of the Nim VM with what I'm trying to do. I'd appreciate any help on this from someone with more experience in macro land.

**Comment by zevv** [comment]
Problems I'm running into when trying to get a grammar 'outside' of a macro:

- quote do / getAst seem not to be able to generate valid AST for a Table[] type, I get `Error: the field 'data' is not accessible`

- Object variants are not properly handled by quote do/getAst, `the fields ...  cannot be initialized together, because they are from conflicting branches in the case object`

**Comment by zevv** [comment]
Well, this turned out to be pretty simple, once I was told how to do it. Thanks for the help, grammar libraries are now implemented and available!

**Comment by zevv** [comment]
The grammar library is in place, which is nice for importing patterns. Next needed would be a way to import a complex library an override only selected parts, for example to add custom captures or code blocks. The idea is that the user can import a grammar and override rules which will shadow the original rules, but any references to the original rules inside that rule will still call the original.

Something like this: import the `uri.URI` grammar, and override the rule `uri.path` from that grammar with a new rule. The new rule addes a capture and a code block, and calls the original rule:

```
peg "line"
  line <- uri.URI
  uri.path <- >uri.path:
    path = $1
```

**Comment by zevv** [comment]
Shadowing implemented in 34239eada84c50bf49f71bd2a69255074eb293ec, closes #9
